### ISSUE TITLE
Better logging for WebGL shader compilation errors

### ISSUE THREAD

**Issue by moribellamy** [godotengine/godot#84428]
This would be helpful in general, and go a ways to help debug #65696

              > Perhaps we are still blocked on needing better instrumentation? Is there a bug open for that?

There's no specific issue open for it; I suggest opening a separate issue to track this.

It looks like there are two components to this:

- GLSL errors are reported by the GLSL shader compiler (which itself is driver-specific), but WebGL fails to report any (look at the above screenshots, it tries to output an empty string below `Vertex shader compilation failed:`).
  - It looks like we can [call `glGetShaderInfoLog()` in WebGL](https://stackoverflow.com/questions/36601815/shaders-are-not-compiling), like in desktop OpenGL. However, this doesn't work currently (this is what Emscripten does for us) so there is likely another issue at hand. In theory, there should be no need to add web-specific code here.
- Logging format is inconsistent with RenderingDevice-based rendering methods (Forward+/Mobile). To resolve this, https://github.com/godotengine/godot/pull/37579 and https://github.com/godotengine/godot/pull/50537 need to be redone for the Compatibility rendering method. (I just checked and this particular issue occurs on desktop too when using Compatibility.)

> Does the Godot steering committee plan to support Web as a first class compilation target for 4.x?

There's no steering committee to speak of. :slightly_smiling_face: 

In open source, people work on what they want to work on, so there's no ETA for resolving this issue. We intend to keep supporting the web export, but we have relatively few contributors available to work on it. The web platform is also prone to a lot of third-party issues that contributors generally struggle dealing with, because they require specific hardware/software configurations (namely, macOS/iOS-only issues).

_Originally posted by @Calinou in https://github.com/godotengine/godot/issues/65696#issuecomment-1793064506_

**Comment by moribellamy** [comment]
I'll take a try at this and report back if I'm able to make progress or not.

**Comment by Calinou** [comment]
I forgot to mention this in the other issue, but we need to make sure this doesn't occur only on macOS (which could hint at an issue with ANGLE on macOS, or macOS' own OpenGL -> Metal translation layer).

Since https://github.com/godotengine/godot/issues/65696#issuecomment-1793064506 doesn't occur on platforms other than macOS, this requires intentionally creating a shader error (e.g. by inserting invalid syntax [here](https://github.com/godotengine/godot/blob/5ee983188de97ae027f9b9c1443438063f708a7e/drivers/gles3/shaders/scene.glsl#L184)), compile an HTML5 export template, use it to export a project then open that project.

**Comment by moribellamy** [comment]
It appears to be more subtle that the experiment @Calinou proposes above.

My experiment:

- inject error into shader code, as recommended
- build godot from source at e38686f85b768a451dc06324fe2471adc8665448
- install emscripten and build the web export template from source 
- export the repro game https://github.com/moribellamy/platonic-game/commit/fb72372b585cb56b61ef509b06bec555028b366e. 
- serve the repro game with `platform/web/serve.py`

The game loads fine in chrome. It does indeed emit the shader error you expect:
![Screenshot 2023-11-10 at 9 19 57 PM](https://github.com/godotengine/godot/assets/13877974/02f55a4e-32df-4618-9550-d4d3f11b2183)

But when you go to slash the slime baddie, the game crashes in the usual way, as reported in https://github.com/godotengine/godot/issues/65696
![Screenshot 2023-11-10 at 9 20 18 PM](https://github.com/godotengine/godot/assets/13877974/a1d2da7b-44fd-4518-a23f-35bddfa386da)

**Comment by moribellamy** [comment]
the build i have cobbled together does at least emit the problematic code per the call `_display_error_with_code(err_string, builder_string);`. i'll see if i cant pull an actual compiler error out of this web

**Comment by moribellamy** [comment]
subscribe [!84741](https://github.com/godotengine/godot/pull/84741)

**Comment by moribellamy** [comment]
I've found nothing. The code which ought to give us a helpful error is here:

https://github.com/godotengine/godot/blob/64150060f89677eaf11229813ae6c5cf8a873802/drivers/gles3/shader_gles3.cpp#L332

Based on my reading and local experimentation, it is a correct query of the OpenGL runtime. I suspect whatever version of WebGL this gets compiled to (by emscripten) is omitting any helpful information about why the vertex is failing to compile.
